File: 0875-koko-eating-bananas/0875-koko-eating-bananas.ts
function minEatingSpeed(piles: number[], h: number): number {
    let ceil = Math.max(...piles)

    let left = 1
    let right = ceil
    while (left <= right) {
        let rate = Math.floor((left+right)/2)

        let total = 0
        for (const pile of piles) {
            total += Math.ceil(pile/rate)
        }
        if (total > h) { //eat too slow 
            left = rate + 1 //eat faster 
        } else {
            right = rate //finish
        }
        if (left == right) return right
    }
    return 0
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        if len(piles) == 1:
            return math.ceil(piles[0] / h)

        left = 1
        right = max(piles)

        while left < right:
            mid = (left + right) // 2 
            
            time = 0
            for pile in piles:
                time += math.ceil(pile / mid)
            print(time, mid, left, right)

            if time > h:
                left = mid + 1
            elif time <= h:
                right = mid

        return right